Implement many-to-one mapping between codes and rules

not-my-profile opened the pull request:
Implements #2186.

charliermarsh commented:
This is really exciting.

I've read through some of the code and commits but not the entirety of the change (yet). From my testing, it seems like every rule has one "primary" code. So `useless-object-inheritance` is mapped to `PLR0205`, `PIE792`, and `UP004`. And no matter how you `--select` it, it's always reported as `PIE792`. Similarly, you have to use `PIE792` to `# noqa` it. I thinks this could be confusing, to `--select UP` and see `PIE792` violations. How do you view it? Is that an accurate description of current behavior?

not-my-profile commented:
> From my testing, it seems like every rule has one "primary" code. I thinks this could be confusing, to --select UP and see PIE792 violations. How do you view it? Is that an accurate description of current behavior?

Yes that's how it currently works. I'd expect us to switch to human-friendly rule names #1773, so no matter which code you used to select in the future ruff will always only report the human-friendly rule name. Until we make that switch the behavior will be a bit confusing, but I don't see a good way around that.

> see the error twice

I am positive that we absolutely do not want to report any error multiple times under any circumstance.

Note that it's totally possible that a rule is enabled via multiple codes, I don't think we want to report multiple codes for a single violation, so we have to pick one ... until we have come up with rule naming guidelines and renamed our rules accordingly.

> should any matching code ignore the violation despite the selector that was used to enable it?

Yes. A rule can be identified via multiple codes. Which code you used to enable a rule doesn't matter.

charliermarsh commented:
I worry that running `--select UP` and seeing a `PIE792` violation is a confusing enough experience that I'd want to really consider whether we enable these at all prior to migrating to human-friendly rule names.

charliermarsh commented:
I'm hesitant to merge this as-is due to some of the confusing behaviors that users will experience around aliasing (e.g., the `--select UP`-to-`PIE` scenario described above).

I don't have great answers for them yet, but I know that if we ship this as-is, it will be confusing for users, and we'll get a lot of feedback and questions stemming from that confusion.

If we want to merge this while giving ourselves time to figure out the best solution for aliasing, what we _could_ do is merge this change but not yet implement any of the actual aliases (apart from, perhaps, deprecating rules, like `BLE001`).

not-my-profile commented:
Right that makes sense. I think the course of action is:

1. Rename our rules to match the naming convention ... this does entail the merging of several rules (see [#2714](https://www.github.com/charliermarsh/ruff/issues/2714)).
   (We do this first since after the next step renaming rules will be a breaking change.)
2. Allow rules to be selected by their name and report violations by their name.
3. Enable the many-to-one mapping.

I already rebased this PR yesterday, which was quite work-intensive since #2583 had landed in between. Further rebasing shouldn't take that much effort, so I'd be alright with leaving this PR open ... but yeah it would be nice to merge this just so that I don't have to deal with merging other changes that may crop up.

I could add an assert statement to the `map_codes` proc macro to assert that at most one code is mapped to one rule ... then we could merge this without any UX changes[^1] and simply remove the assert statements once we get to the above mentioned 3rd step.

[^1]: Aside from the C, C9, T, T1, T2 prefix deprecations in the 7th commit.

sbrugman commented:
(Heads-up: I will rename the pathlib rules in https://github.com/charliermarsh/ruff/pull/2348 to os-path - the thing we detect)

not-my-profile commented:
I think that for landing this PR, I'll need some coordination that this will be merged next, since this PR changes the format of the frequently edited `registry.rs` file.

charliermarsh commented:
👍 I’m happy to merge this next assuming it doesn’t change the UX right now — is that the case? I have to re-read the code but I’ll review tomorrow after the JetBrains webinar, and hopefully can merge tomorrow afternoon or evening.

not-my-profile commented:
Ok great. Yes after I have dropped the last commit demonstrating the mapping and added the aforementioned assert statement there shouldn't be any UX changes.

charliermarsh commented:
(Returning to this now.)

not-my-profile commented:
Let me know if you have questions.

charliermarsh commented on ruff.schema.json:
```diff
@@ -1396,7 +1395,6 @@
         "C415",
         "C416",
         "C417",
-        "C9",
```
So these work on the CLI (based on my testing), what's the rationale for excluding them?

not-my-profile replied:
They only continue to work because I added redirects, see the "Update JSON schema" commit.

They are no longer generated by `cargo dev generate-json-schema` since these are partial linter prefixes ... I'd say we should consider them deprecated.

charliermarsh replied:
Is this not the McCabe-equivalent of (e.g.) `--select B` for `flake8-bugbear`?

charliermarsh replied:
(As an aside, would be nice to just remove these conflicts at some point.)

not-my-profile replied:
> Is this not the McCabe-equivalent of (e.g.) --select B for flake8-bugbear?

No as per `#[prefix = "C90"] McCabe` in `registry.rs` the equivalent is C90.

not-my-profile replied:
Note that this is also consistent with the upstream mccabe:

https://github.com/PyCQA/mccabe/blob/master/setup.py#L38

charliermarsh commented on crates/ruff_cli/src/printer.rs:
```diff
@@ -601,7 +602,7 @@ fn print_message<T: Write>(
                 line_start: message.location.row(),
                 annotations: vec![SourceAnnotation {
-                    label: message.kind.rule().code(),
+                    label: &label,
```
Removed from inline just for clarity? Or some other reason?

not-my-profile replied:
Removed from inline due to borrow checking reasons. Otherwise borrowchk complains due to returning a temporary reference to an owned value.

charliermarsh commented on crates/ruff/src/settings/pyproject.rs:
```diff
@@ -225,8 +225,8 @@ ignore = ["E501"]
                 ruff: Some(Options {
-                    extend_select: Some(vec![RuleCodePrefix::RUF100.into()]),
-                    ignore: Some(vec![RuleCodePrefix::E501.into()]),
+                    extend_select: Some(vec![codes::Ruff::_100.into()]),
```
So this is a result of the prefix now being encoded via `codes::Ruff`, and so the identifier itself starts with a number, and so we need the underscore. Is that correct?

not-my-profile replied:
Yes, see `get_prefix_ident` in `ruff_macros/src/rule_code_prefix.rs`:

> // Identifiers in Rust may not start with a number.

charliermarsh commented:
Cool this looks good to me -- merging...
